src/shred_collector/shred_tracker.zig

```diff
@@ -0,0 +1,240 @@
+        return .{
+            .start_slot = slot,
+            .current_bottom_slot = slot orelse 0,
+            .logger = logger,
+        };
+    }
+
+    pub fn maybeSetStart(self: *Self, start_slot: Slot) void {
+        if (self.start_slot == null) {
+            self.start_slot = start_slot;
+            self.current_bottom_slot = start_slot;
+        }
+    }
+
+    pub fn registerShred(
+        self: *Self,
+        slot: Slot,
+        shred_index: u64,
+    ) !void {
+        self.mux.lock();
+        defer self.mux.unlock();
+
+        self.maybeSetStart(slot);
+        self.max_slot_seen = @max(self.max_slot_seen, slot);
+        const monitored_slot = try self.getSlot(slot);
+        const new = try monitored_slot.record(shred_index);
+        if (new) self.logger.debugf("new slot: {}", .{slot});
+        self.max_slot_processed = @max(self.max_slot_processed, slot);
+    }
+
+    pub fn setLastShred(self: *Self, slot: Slot, index: usize) !void {
+        self.mux.lock();
+        defer self.mux.unlock();
+
+        self.maybeSetStart(slot);
+        const monitored_slot = try self.getSlot(slot);
+        if (monitored_slot.last_shred) |old_last| {
+            monitored_slot.last_shred = @min(old_last, index);
+        } else {
+            monitored_slot.last_shred = index;
+        }
+    }
+
+    /// returns whether it makes sense to send any repair requests
+    pub fn identifyMissing(self: *Self, slot_reports: *MultiSlotReport) !bool {
+        if (self.start_slot == null) return false;
+        self.mux.lock();
+        defer self.mux.unlock();
+
+        var found_an_incomplete_slot = false;
+        slot_reports.clearRetainingCapacity();
+        const timestamp = std.time.milliTimestamp();
+        const last_slot_to_check = @max(self.max_slot_processed, self.current_bottom_slot);
+        for (self.current_bottom_slot..last_slot_to_check + 1) |slot| {
+            const monitored_slot = try self.getSlot(slot);
+            if (monitored_slot.first_received_timestamp_ms + MIN_SLOT_AGE_TO_REPORT_AS_MISSING > timestamp) {
+                continue;
+            }
+            var slot_report = try slot_reports.addOne();
+            slot_report.slot = slot;
+            try monitored_slot.identifyMissing(&slot_report.missing_shreds);
+            if (slot_report.missing_shreds.items.len > 0) {
+                found_an_incomplete_slot = true;
+            } else {
+                slot_reports.drop(1);
+            }
+            if (!found_an_incomplete_slot) {
+                self.logger.debugf("finished slot: {}", .{slot}); // FIXME not always logged
+                self.current_bottom_slot = @max(self.current_bottom_slot, slot + 1);
+                monitored_slot.* = .{};
+            }
+        }
+        return true;
+    }
+
+    fn getSlot(self: *Self, slot: Slot) error{ SlotUnderflow, SlotOverflow }!*MonitoredSlot {
+        if (slot > self.current_bottom_slot + num_slots - 1) {
```

Comment:
Good point. Slots can be skipped.

At one point my mental model was that skipped slots were implemented by having the leader of the next slot fill in the gap with an empty block. But that would fail signature verification, so that can't be right.

I think I need to add some logic to check the parent offset in data shreds to identify skipped slots.

Reply:
i was thinking even simpler where you request shreds 1-20 but you receive them out of order

Reply:
Yeah it's just going to request 5000 later once it progresses to that point. There's definitely a better way of doing this but the shred tracker was already a pretty big time sink considering it's going to be replaced with blockstore.